File: src/main/java/com/epri/dss/shared/impl/PStackImpl.java
package com.epri.dss.shared.impl;

import com.epri.dss.shared.PStack;

public class PStackImpl<T> extends StackBaseImpl implements PStack<T> {

	private T[] items;

	public PStackImpl(int initSize) {
		super(initSize);
		items = (T[]) new Object[initSize];
	}

	public void push(T p) {
		T[] newItems;
		int size, l;

		numItems += 1;
		if (numItems > maxItems) {
			// resize array
			maxItems += increment;
			newItems = (T[]) new Object[maxItems];
			size = items.length;
			l = Math.min(size, maxItems);
			if (l > 0)
				System.arraycopy(items, 0, newItems, 0, l);
			items = newItems;
		}
		items[numItems] = p;
	}

	public T pop() {
		T result;
		if (numItems > 0) {
			result = items[numItems - 1];
			numItems -= 1;
		} else {
			result = null;
		}
		return result;
	}

}

File: pydss/shared/PStackImpl.py
from dss.shared.impl.StackBaseImpl import (StackBaseImpl,)
from dss.common.impl.Utilities import (Utilities,)
from dss.shared.PStack import (PStack,)


class PStackImpl(StackBaseImpl, PStack):
    items = None

    def __init__(self, initSize):
        super(PStackImpl, self)(initSize)
        self.items = [None] * initSize

    def push(self, p):
        self.numItems += 1
        if self.numItems > self.maxItems:
            self.maxItems += self.increment
            self.items = Utilities.resizeArray(self.items, self.maxItems)
        self.items[self.numItems] = p

    def pop(self):
        if self.numItems > 0:
            Result = self.items[self.numItems - 1]
            self.numItems -= 1
        else:
            Result = None
        return Result
